### PR TITLE
feat[next][dace]: Fix lowering of nested let-statements

### PR DESCRIPTION
**Pull request by edopao**
This PR fixes one corner case of nested let-statements, discovered in `test_tuple_unpacking_star_multi` during GTIR integration. Test case added.

Additionally, fixed handling of symbol already defined in SDFG for #1695.

### CODE REVIEW

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py [hunk 1]
```diff
@@ -258,18 +258,20 @@ def _add_storage(
...
-                # storage for field arguments. We assume that the scalar argument
-                # for field size, if present, always follows the field argument.
                 if name in sdfg.symbols:
-                    assert sdfg.symbols[name].dc_dtype == dc_dtype
+                    # Sometimes, when the field domain is implicitly derived from the
+                    # field domain, the gt4py lowering adds the field size as a scalar
+                    # argument to the program IR. Suppose a field '__sym', then gt4py
+                    # will add '__sym_size_0'.
+                    # Therefore, here we check whether the shape symbol was already
+                    # created by `_make_array_shape_and_strides`, when allocating
```
Comment (philip-paul-mueller) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py:267]:
```suggestion
                    # created by `_make_array_shape_and_strides()`, when allocating
```
OPTIONAL: I like this more to make clear that it is a function.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py [hunk 2]
```diff
@@ -649,65 +651,56 @@ def _flatten_tuples(
...
-            nstate.degree(output_data.dc_node) == 0
-            for output_data in lambda_output_data
-            if output_data.dc_node.data in input_memlets
-        )
-        # keep only non-isolated output nodes
         lambda_outputs = {
             output_data.dc_node.data
             for output_data in lambda_output_data
-            if output_data.dc_node.data not in input_memlets
+            if output_data.dc_node.desc(nsdfg).transient
```
Comment (philip-paul-mueller) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py:657]: Data inside the nested SDFG that serves as input/output can not be transient.
This does not make sense to me.
If it has to do something with `make_temps()` then please indicate this as it is very strange.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py [hunk 3]
```diff
@@ -649,65 +651,56 @@ def _flatten_tuples(
...
+        for connector, memlet in input_memlets.items():
+            if connector in lambda_arg_nodes:
+                src_node = lambda_arg_nodes[connector].dc_node
+            else:
+                src_node = head_state.add_access(memlet.data)
 
-                head_state.add_edge(src_node, None, nsdfg_node, connector, memlet)
+            head_state.add_edge(src_node, None, nsdfg_node, connector, memlet)
 
         def make_temps(
```
Comment (philip-paul-mueller) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py:677]: Could you add a short description what this function does.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py [hunk 4]
```diff
@@ -649,65 +651,56 @@ def _flatten_tuples(
...
             else:
-                nstate.remove_node(output_data.dc_node)
-                data_node = head_state.add_access(output_data.dc_node.data)
-                return gtir_builtin_translators.FieldopData(
-                    data_node, output_data.gt_dtype, output_data.local_offset
+                dc_node = head_state.add_access(output_data.dc_node.data)
+                temp_field = gtir_builtin_translators.FieldopData(
+                    dc_node, output_data.gt_dtype, output_data.local_offset
                 )
+            # isolated access node will make validation fail
```
Comment (philip-paul-mueller) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py:700]: Why should this happen.
If the node is isolated, this would mean that the output is not needed in the first place or do I miss something.

**Review by philip-paul-mueller — COMMENTED** [review]
There are some points that needs clarification.

**Review by philip-paul-mueller — APPROVED** [review]
For symmetry I would keep the parenthesis on the true case too.
But this is a matter of style.
LGTM.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py [hunk 5]
```diff
@@ -648,9 +651,21 @@ def _flatten_tuples(
 
         # Process lambda outputs
         #
+        # The output arguments do not really exist, so they are not allocated before
+        # visiting the lambda expression. Therefore, the result appears inside the
+        # nested SDFG as transient array/scalar storage. The exception is given by
+        # input arguments that are just passed through and returned by the lambda,
+        # e.g. when the lambda is constructing a tuple: in this case, the result
+        # data is non-transient, because it corresponds to an input node.
```
Comment (philip-paul-mueller) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py:659]: I would add that this is corrected by `make_temps()`.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py [hunk 6]
```diff
@@ -677,6 +692,21 @@ def _flatten_tuples(
...
+            1st if-branch:
+            Transient nodes actually contain some result produced by the dataflow
+            itself, therefore these nodes are changed to non-transient and an output
+            edge will write the result from the nested SDFG to a new intermediate
+            data node in the parent context.
+
+            2nd and 3rd if-branch:
+            Non-transient nodes are just input nodes that are immediately returned
+            by the lambda expression. Therefore, these nodes are already available
+            in the parent context and can be directly accessed there.
```
Comment (philip-paul-mueller) [inline src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py:708]: I would put this description as comments.
Because, a docstring is supposed to be viewed without the source code.

**Review by philip-paul-mueller — APPROVED** [review]
There are two suggestions but it looks good.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py (hunk 5 above)
Comment (edopao) [reply src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py:652]: Done.

File: src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py (hunk 6 above)
Comment (edopao) [reply src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py]: Good point.
